include/ic_gameobject.h

```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include "ic_components.h"
+
+#include <memory>
+#include <typeindex>
+#include <typeinfo>
+#include <vector>
+
+namespace IC {
+    class GameObject {
+    public:
+        GameObject(std::string name);
+        ~GameObject();
+
+        std::shared_ptr<Transform> &GetTransform() { return _transform; }
+
+        template <typename T> std::shared_ptr<T> AddComponent() {
+            _componentMap[typeid(T)] = std::make_shared<T>();
+            return static_pointer_cast<T>(_componentMap[typeid(T)]);
+        }
+
+        template <typename T> bool HasComponent() { return _componentMap.contains(typeid(T)); }
+
+        template <typename T> std::shared_ptr<T> GetComponent() {
+            if (HasComponent<T>()) {
```

Comment:
is it ok to not return by reference here and in `AddComponent()`? Was getting this error `error: cannot bind non-const lvalue reference of type ‘std::shared_ptr<IC::DirectionalLight>&’ to an rvalue of type ‘std::shared_ptr<IC::DirectionalLight>’`

Reply:
Yes this is good. It's better to just return the pointer and not a reference because the function can fail and return nullptr (which is where the error was coming from).